src/core/services/theming/theming.js

```diff
@@ -639,13 +680,42 @@ function parseRules(theme, colorType, rules) {
       hue = theme.colors[colorType].hues[hue];
     }
 
-    return rgba( (PALETTES[ theme.colors[colorType].name ][hue] || '')[contrast ? 'contrast' : 'value'], opacity );
+    var colorDetails = (PALETTES[ theme.colors[colorType].name ][hue] || '');
+
+    // If user has specified a foreground color, use those
+    if (colorType === 'background' && contrast && regexColorType !== 'foreground' && colorDetails.contrastType == defaultBgContrastType) {
+      // Don't process if colorType was changed
+      switch (opacity) {
+        default:
+          if (theme.foregroundPalette['1']) return rgba( colorToRgbaArray( theme.foregroundPalette['1'] ) );
+          break;
+        case 'secondary':
+        case 'icon':
+          if (theme.foregroundPalette['2']) return rgba( colorToRgbaArray( theme.foregroundPalette['2'] ) );
```

Comment:
the line feels overloaded, put in brackets or at least break return to a separate line